File: src/main/java/jugglinglab/notation/MHNSymmetry.java
//
// MHNSymmetry.java
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.notation;

import jugglinglab.util.*;

public class MHNSymmetry {
  int type;
  int numjugglers;
  Permutation jugglerperm;
  int delay = -1;

  public static final int TYPE_DELAY = 1; // types of symmetries
  public static final int TYPE_SWITCH = 2;
  public static final int TYPE_SWITCHDELAY = 3;

  public MHNSymmetry(int type, int numjugglers, String jugperm, int delay)
      throws JuggleExceptionInternal {
    setType(type);
    setJugglerPerm(numjugglers, jugperm);
    setDelay(delay);
  }

  public int getType() {
    return type;
  }

  protected void setType(int type) {
    this.type = type;
  }

  public int getNumberOfJugglers() {
    return numjugglers;
  }

  public Permutation getJugglerPerm() {
    return jugglerperm;
  }

  protected void setJugglerPerm(int nj, String jp) throws JuggleExceptionInternal {
    numjugglers = nj;
    try {
      if (jp == null) jugglerperm = new Permutation(numjugglers, true);
      else jugglerperm = new Permutation(numjugglers, jp, true);
    } catch (JuggleException je) {
      throw new JuggleExceptionInternal(je.getMessage());
    }
  }

  public int getDelay() {
    return delay;
  }

  protected void setDelay(int del) {
    delay = del;
  }
}

File: composeApp/src/commonMain/kotlin/jugglinglab/notation/MhnSymmetry.kt
//
// MhnSymmetry.kt
//
// Copyright 2002-2025 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.notation

import jugglinglab.util.Permutation

data class MhnSymmetry(
    val type: Int,
    val numberOfJugglers: Int,
    val jugPerm: String?,
    val delay: Int
) {
    val jugglerPerm: Permutation = if (jugPerm == null) {
            Permutation(numberOfJugglers, reverses = true)
        } else {
            Permutation(numberOfJugglers, jugPerm, reverses = true)
        }

    companion object {
        const val TYPE_DELAY: Int = 1  // types of symmetries
        const val TYPE_SWITCH: Int = 2
        const val TYPE_SWITCHDELAY: Int = 3
    }
}
